packages/normalizr/src/denormalize.ts

```diff
@@ -21,72 +32,131 @@ const unvisitEntity = (
+  entities: Record<string, Record<string, any>>,
+  entityCache: DenormalizeCache['entities'],
+  resultCache: WeakListMap<object, any>,
+) => {
+  const localCache = {};
   const getEntity = getEntities(entities);
 
-  return [
-    function unvisit(input: any, schema: any): [any, boolean, boolean] {
-      if (!schema) return [input, true, false];
-
-      if (!schema.denormalize || typeof schema.denormalize !== 'function') {
-        if (typeof schema === 'function') {
-          if (input instanceof schema) return [input, true, false];
-          return [new schema(input), true, false];
-        } else if (typeof schema === 'object') {
-          const method = Array.isArray(schema)
-            ? arrayDenormalize
-            : objectDenormalize;
-          return method(schema, input, unvisit);
-        }
+  function unvisit(
+    input: any,
+    schema: any,
+  ): [denormalized: any, found: boolean, deleted: boolean] {
+    if (!schema) return [input, true, false];
+
+    if (!schema.denormalize || typeof schema.denormalize !== 'function') {
+      if (typeof schema === 'function') {
+        if (input instanceof schema) return [input, true, false];
+        return [new schema(input), true, false];
+      } else if (typeof schema === 'object') {
+        const method = Array.isArray(schema)
+          ? arrayDenormalize
+          : objectDenormalize;
+        return method(schema, input, wrappedUnvisit);
       }
+    }
 
-      // null is considered intentional, thus always 'found' as true
-      if (input === null) {
-        return [input, true, false];
-      }
+    // null is considered intentional, thus always 'found' as true
+    if (input === null) {
+      return [input, true, false];
+    }
 
-      if (isEntity(schema)) {
-        // unvisitEntity just can't handle undefined
-        if (input === undefined) {
-          return [input, false, false];
-        }
-        return unvisitEntity(input, schema, unvisit, getEntity, cache);
+    if (isEntity(schema)) {
+      // unvisitEntity just can't handle undefined
+      if (input === undefined) {
+        return [input, false, false];
       }
+      return unvisitEntity(
+        input,
+        schema,
+        wrappedUnvisit,
+        getEntity,
+        localCache,
+        entityCache,
+      );
+    }
 
-      if (typeof schema.denormalize === 'function') {
-        return schema.denormalize(input, unvisit);
-      }
+    if (typeof schema.denormalize === 'function') {
+      return schema.denormalize(input, wrappedUnvisit);
+    }
 
-      return [input, true, false];
-    },
-    cache,
-  ] as const;
+    return [input, true, false];
+  }
+
+  // make unvisit that tracks entity list here
+  const globalKey: EntityInterface[] = [];
+  const wrappedUnvisit = (input: any, schema: any) => {
+    const ret: [any, boolean, boolean] = unvisit(input, schema);
+    // pass over undefined in key
+    if (ret[0] && schema && isEntity(schema)) globalKey.push(ret[0]);
```

Comment:
`ret[0]` here has a type of `any`, and is then pushed into `globalKey`, which means it should have a type of `EntityInterface`. What's the reason not to do:
```suggestion
    const ret: [EntityInterface, boolean, boolean] = unvisit(input, schema);
    // pass over undefined in key
    if (ret[0] && schema && isEntity(schema)) globalKey.push(ret[0]);
```
?

Reply:
isEntity(schema) determines whether it should be pushed. otherwise it is not expected to be EntityInterface.